### PR TITLE
[NHWC] Added Asm iGemm WrW. Some rework in Asm iGemm Fwd/Bwd.

### PR DESCRIPTION
**Pull request by shaojiewang**
This PR aims to add asm igemm for wrw pass, layout NHWC:
- [x] fp32
- [x] fp16

Performance test for wrw: [perf_data_vs_nchw](https://amdcloud-my.sharepoint.com/:x:/g/personal/shaowang_amd_com/EeagiOv0w0dKq1o91slvkGYByMLUjA0P7Wt8r88HeCPw5w?e=sMoifx)
resnet50 vs cudnn: [resnet50](https://amdcloud-my.sharepoint.com/:x:/g/personal/shaowang_amd_com/ERTdUMxTFgxOrBkQ17wy__4Bx3WPtSzjAmjdcnBDuRbewA?e=lMb3RW)
All perf data on MI100 is collected by MIOpenDriver.

To avoid string compare in solver code, the precision attribute in struct `PerformanceConfigAsmImplicitGemmGTC` is also re-typed by `miopenDataType_t`. So some fwd and bwd files are also changed.

Besides the asm files which are generated by igemmGen, the following cpp or cmake files are also changed:
```
src/CMakeLists.txt
src/include/miopen/conv/asm_implicit_gemm.hpp
src/include/miopen/solver.hpp
src/mlo_dir_conv.cpp
src/solver.cpp
src/solver/conv_asm_implicit_gemm_gtc_bwd_nhwc.cpp
src/solver/conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp
src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp
src/solver/conv_asm_implicit_gemm_gtc_perf_config.cpp
test/CMakeLists.txt
```

### CODE REVIEW

**Comment by atamazov** [comment]
@shaojiewang Please 
- Resolve merge conflicts
- Assign reviewers
- Assign urgency and value label, plus milestone
- Attach results of performance testing (baseline develop vs this PR)

Thanks!

**Comment by carlushuang** [comment]
@shaojiewang 
pls merge conflict first

**Comment by shaojiewang** [comment]
> @shaojiewang
> pls merge conflict first

yes, still working on that.

**Comment by atamazov** [comment]
Formatting issues.

**Comment by atamazov** [comment]
test_conv_igemm_dynamic_xdlops_nhwc_wrw failed on CI.

Please merge from develop -- we just migrated from ROCm 3.7 to 4.2.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -1076,20 +1076,20 @@ COMMAND ${DYNAMIC_IMPLICITGEMM_XDLOPS_NHWC_BWD_ENVS} $<TARGET_FILE:test_conv2d>
...
-COMMAND ${DYNAMIC_IMPLICITGEMM_XDLOPS_NHWC_WRW_ENVS} $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --verbose --input  64  32 73 73 --weights  64  32 3 3 --pads_strides_dilations 1 1 1 1 1 1 --disable-forward --disable-backward-weights --in_layout NHWC --fil_layout NHWC --out_layout NHWC
-COMMAND ${DYNAMIC_IMPLICITGEMM_XDLOPS_NHWC_WRW_ENVS} $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --verbose --input  16  64 56 56 --weights  64  64 3 3 --pads_strides_dilations 1 1 1 1 1 1 --disable-forward --disable-backward-weights --in_layout NHWC --fil_layout NHWC --out_layout NHWC
-COMMAND ${DYNAMIC_IMPLICITGEMM_XDLOPS_NHWC_WRW_ENVS} $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --verbose --input  16  16 25 25 --weights  64  16 3 3 --pads_strides_dilations 0 0 1 1 1 1 --disable-forward --disable-backward-weights --in_layout NHWC --fil_layout NHWC --out_layout NHWC
+COMMAND ${DYNAMIC_IMPLICITGEMM_XDLOPS_NHWC_WRW_ENVS} $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --verbose --input  64 256  7  7 --weights 128 256 1 1 --pads_strides_dilations 0 0 1 1 1 1 --disable-forward --disable-backward-data --in_layout NHWC --fil_layout NHWC --out_layout NHWC
```
Comment (atamazov) [inline test/CMakeLists.txt:1079]: [Recommendation] You may wish to borrow approach from `ARGS_REGRESSION_ISSUE_1012`, see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1013/commits/b6a3ff73ff20094082869d3692974d33c2b11f9a#r662418510

**Review by atamazov — COMMENTED** [review]
This can be merged as is, but in this case all review comments must be resolved in separate PR. Please let me know if you would like go this way.

File: test/CMakeLists.txt (hunk 1 above)
Comment (shaojiewang) [reply test/CMakeLists.txt:1079]: OK, Thanks.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+#include <miopen/generic_search.hpp>
+#include <miopen/conv/wrw_invoke_params.hpp>
+#include <miopen/solver/implicitgemm_util.hpp>
+#include <miopen/gcn_asm_utils.hpp>
+#include <miopen/tensor_ops.hpp>
+#include <miopen/conv/asm_implicit_gemm.hpp>
+
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_WRW_GTC_XDLOPS_NHWC)
+
+#define WRW_MAX_GEMM_K_SPLITS 10
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:39]: It is the range for tuning case. Normally, the best tuning param of gemmk splits is less than 10. So, I put a 10 here.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 3]
```diff
@@ -123,6 +123,22 @@ static inline size_t ComputeLog2GemmKGlobalSplitsWith2DMerge(size_t current_grid
...
 }
 
+// calculate gemm_k_global_splits
+// with assumption that some dimensions will merge into a single dimension
+static inline size_t ComputeGemmKGlobalSplitsWith2DMerge(size_t current_grid_size,
+                                                         // size_t merge_dimension,
+                                                         // size_t gemm_k_per_block,
+                                                         size_t occupancy)
+{
+    size_t num_cu               = 120;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:133]: OK. done in this PR.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+
+    vgpr_usage = tensor_a_thread_lengths[1] * tensor_a_thread_lengths[3] / a_elements_per_vgpr +
+                 tensor_b_thread_lengths[1] * tensor_b_thread_lengths[3] / b_elements_per_vgpr +
+                 tensor_a_thread_lengths[1] * tensor_a_thread_lengths[3] / sz_per_element +
+                 tensor_b_thread_lengths[1] * tensor_b_thread_lengths[3] / sz_per_element +
+                 aux_vgpr_usage;
+    if(GetTypeSize(precision) == 2)
+    {
+        if(lds_single >= 32 * 1024 || (lds_single <= 16 * 1024 && lds_single > 8 * 1024 &&
+                                       acc_usage < 128 && vgpr_usage <= 84))
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:248]: > Please explain the meaning of these magic numbers.

OK

> IIRC LDS size is 64K, right?

Right. To keep occupancy bigger than 1, we do not want to use more than half LDS, which is 32K.

**Comment by shaojiewang** [comment]
> This can be merged as is, but in this case all review comments must be resolved in separate PR. Please let me know if you would like go this way.

I can make these changes within this PR.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+            lds_usage = lds_single * 2;
+        }
+    }
+    else
+    {
+        lds_usage = lds_single;
+    }
+
+    // std::cout << "lds_usage=" << lds_usage << ", acc_usage=" << acc_usage << ", vgpr_usage=" <<
+    // vgpr_usage << std::endl;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:264]: Yes, Thanks.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 6]
```diff
@@ -0,0 +1,842 @@
...
+    kernel.l_wk.clear();
+    kernel.l_wk.push_back(block_size);
+    kernel.l_wk.push_back(1);
+    kernel.l_wk.push_back(1);
+
+    GenerateClangDefsym(options, "ROCM_METADATA_VERSION", ctx.rmv.UseV3() ? 5 : 4);
+
+    kernel.comp_options = options.str();
+
+    MIOPEN_LOG_I2("ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC: " + config.ToString());
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:743]: Not necessary, just to check config param. So I use a MIOPEN_LOG_T here.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 7]
```diff
@@ -0,0 +1,842 @@
...
+    }
+    else
+    {
+        lds_usage = lds_single;
+    }
+
+    // std::cout << "lds_usage=" << lds_usage << ", acc_usage=" << acc_usage << ", vgpr_usage=" <<
+    // vgpr_usage << std::endl;
+
+    occupancy = std::min(64 * 1024 / lds_usage, std::min(256 / acc_usage, 256 / vgpr_usage));
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: changed

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 8]
```diff
@@ -0,0 +1,842 @@
...
+                min_pad_pixel  = cur_pad_pixel;
+                selected_index = i;
+            }
+        }
+
+        size_t current_grid_size;
+        size_t occupancy;
+        std::tie(std::ignore, std::ignore, current_grid_size, occupancy) =
+            GetImplicitGemmGtcDynamicWrwXdlopsNHWCKernel(ctx, config_list[selected_index]);
+        bool need_k_split = current_grid_size <= 600;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: ok

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 9]
```diff
@@ -0,0 +1,842 @@
...
+                continue;
+
+            if(m_per_block == config.gemm_m_per_block && n_per_block == config.gemm_n_per_block &&
+               k_per_block == config.gemm_k_per_block)
+            {
+                size_t current_grid_size;
+                size_t occupancy;
+                std::tie(std::ignore, std::ignore, current_grid_size, occupancy) =
+                    GetImplicitGemmGtcDynamicWrwXdlopsNHWCKernel(ctx, config);
+                bool need_k_split = current_grid_size <= 600;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: ok

**Comment by junliume** [comment]
@shaojiewang @atamazov it seems that test_conv_igemm_mlir_xdlops test has failed via "Cannot contact pytorch-vg-20-1.amd.com". Maybe resource offline? @okakarpa Thanks!

**Comment by shaojiewang** [comment]
> @shaojiewang @atamazov it seems that test_conv_igemm_mlir_xdlops test has failed via "Cannot contact pytorch-vg-20-1.amd.com". Maybe resource offline? @okakarpa Thanks!

Test of PR #957 is also blocked by the same issue.

**Comment by junliume** [comment]
> > @shaojiewang @atamazov it seems that test_conv_igemm_mlir_xdlops test has failed via "Cannot contact pytorch-vg-20-1.amd.com". Maybe resource offline? @okakarpa Thanks!
> 
> Test of PR #957 is also blocked by the same issue.

@shaojiewang there are other jobs later succeeded so a hot bypass is to restart the job. We need to look into these vg nodes.

**Review by junliume — APPROVED** [review]

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 10]
```diff
@@ -418,8 +427,8 @@ void PerformanceConfigAsmImplicitGemmGTCWrwXdlopsNHWC::HeuristicInit(const Convo
         size_t occupancy;
         std::tie(std::ignore, std::ignore, current_grid_size, occupancy) =
             GetImplicitGemmGtcDynamicWrwXdlopsNHWCKernel(ctx, config_list[selected_index]);
-        bool need_k_split = current_grid_size <= 600;
-        size_t gks        = ComputeGemmKGlobalSplitsWith2DMerge(current_grid_size, occupancy);
+        bool need_k_split = current_grid_size <= non_split_gridsize;
+        size_t gks = ComputeGemmKGlobalSplitsWith2DMerge(current_grid_size, occupancy, NUM_CUS);
```
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:431]:
```suggestion
        const auto num_cu = ctx.GetStream().GetMaxComputeUnits();
        size_t gks = ComputeGemmKGlobalSplitsWith2DMerge(current_grid_size, occupancy, num_cu);
```

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 11]
```diff
@@ -443,13 +452,11 @@ void PerformanceConfigAsmImplicitGemmGTCWrwXdlopsNHWC::HeuristicInit(const Convo
                 size_t occupancy;
                 std::tie(std::ignore, std::ignore, current_grid_size, occupancy) =
                     GetImplicitGemmGtcDynamicWrwXdlopsNHWCKernel(ctx, config);
-                bool need_k_split = current_grid_size <= 600;
-                size_t gks = ComputeGemmKGlobalSplitsWith2DMerge(current_grid_size, occupancy);
+                bool need_k_split = current_grid_size <= non_split_gridsize;
+                size_t gks =
+                    ComputeGemmKGlobalSplitsWith2DMerge(current_grid_size, occupancy, NUM_CUS);
```
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:457]:
```suggestion
                    ComputeGemmKGlobalSplitsWith2DMerge(current_grid_size, occupancy, num_cu);
```

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 12]
```diff
@@ -705,7 +703,7 @@ ConvSolution ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC::GetSolution(
 
     size_t gemm_k_global_splits =
         config.gemm_k_global_split >= 1
-            ? ComputeGemmKGlobalSplitsWith2DMerge(grid_size, config.gemm_k_global_split)
+            ? ComputeGemmKGlobalSplitsWith2DMerge(grid_size, config.gemm_k_global_split, NUM_CUS)
```
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:706]:
```suggestion
            ? ComputeGemmKGlobalSplitsWith2DMerge(grid_size, config.gemm_k_global_split, ctx.GetStream().GetMaxComputeUnits();)
```

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 13]
```diff
@@ -37,7 +37,7 @@
 MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_WRW_GTC_XDLOPS_NHWC)
 
 #define WRW_MAX_GEMM_K_SPLITS 10
-//#define DEBUG_IGEMM_ASM_WRW_NHWC_CHECK_VALID_TILE_LIST
+#define NUM_CUS 120
```
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:40]: Please remove this.

**Review by atamazov — COMMENTED** [review]
This can be merged as is, but in this case all review comments must be resolved in separate PR. Please let me know if you would like go this way.

File: src/include/miopen/conv/asm_implicit_gemm.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/conv/asm_implicit_gemm.hpp]: Not completely, please see my next review, thanks.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp (hunk 13 above)
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: removed.

**Review by atamazov — DISMISSED** [review]
LGTM!

**Comment by shaojiewang** [comment]
CI tests have been passed.

**Review by atamazov — APPROVED** [review]
I fixed merge conflicts, so approving again.

**Pull request merged**